labscript_devices/AD9959DDSSweeper/labscript_devices.py

```diff
@@ -0,0 +1,262 @@
+
+        This labscript device provides up to four channels of DDS outputs. It is designed to be connected to a pseudoclock clockline.
+
+        Args:
+            name (str): python variable name to assign to the AD9959DDSSweeper
+            parent_device (:class:`~.ClockLine`):
+                Pseudoclock clockline used to clock DDS parameter changes.
+            com_port (str): COM port assigned to the AD9959DDSSweeper by the OS.
+                On Windows, takes the form of `COMd` where `d` is an integer.
+            pico_board (str): The version of pico board used, pico1 or pico2.
+            sweep_mode (int):
+                The DDS Sweeper firmware can set the DDS outputs in either fixed steps or sweeps of the amplitude, frequency, or phase.
+                At this time, only steps are supported, so sweep_mode must be 0.
+            ref_clock_external (int): Set to 0 to have Pi Pico provide the reference clock to the AD9959 eval board. Set to 1 for another source of reference clock for the AD9959 eval board.
+            ref_clock_frequency (float): Frequency of the reference clock. If ref_clock_external is 0, the Pi Pico system clock will be set to this frequency. If the PLL is used, ref_clock_frequency * pll_mult must be between 100 MHz and 500 MHz. If the PLL is not used, ref_clock_frequency must be less than 500 MHz.
+            pll_mult: the AD9959 has a PLL to multiply the reference clock frequency. Allowed values are 1 or 4-20.
+        '''
+        IntermediateDevice.__init__(self, name, parent_device, **kwargs)
+        self.BLACS_connection = '%s' % com_port
+        
+        if pico_board in self.allowed_boards:
+            self.pico_board = pico_board
+        else:
+            raise LabscriptError(f'Pico board specified not in {self.allowed_boards}')
+        
+        # store mode data
+        self.sweep_mode = sweep_mode
+        self.ref_clock_frequency = ref_clock_frequency
+        # Check clocking
+        if ref_clock_frequency * pll_mult > 500e6:
+            raise ValueError('DDS system clock frequency must be less than 500 MHz')
+        elif pll_mult > 1 and ref_clock_frequency * pll_mult < 100e6:
+            raise ValueError('DDS system clock frequency must be greater than 100 MHz when using PLL')
+        elif not ref_clock_external and ref_clock_frequency > 133e6:
+            raise ValueError('ref_clock_frequency must be less than 133 MHz when clock is provided by Pi Pico')
+
+        self.dds_clock = ref_clock_frequency * pll_mult
+        self.clk_scale = 2**32 / self.dds_clock
+
+    @property
+    def clock_limit(self):
+        '''Dynamically computs clock limit based off of number of dynamic 
+        channels and reference clock frequency.'''
+        num_dyn_chans = sum(isinstance(child, DDS) for child in self.child_devices)
+        if num_dyn_chans == 0:
+            # Set to worst case 
+            # 4 channels, step mode, default 125 MHz pico ref clk
+            return 100000
```

Comment:
This is an interesting case, since the clock_limit in this situation is immaterial since there will be no dynamic channels asking for triggers anyway. It might be good to check that you can compile a shot with a high sample rate DAQ ramp using the same psuedoclock and no dynamic outputs from the DDS. There should be no complaint. Adding a dynamic DDS doing something at the same time as the DAQ should cause an error.

Reply:
Based on a read through I did of the `clock_limit` processing, I am a bit concerned this might not work.
As far as I can tell, the `ClockLine` (parent device) will check `clock_limit` when the `IntermediateDevice` (the `AD9959DDSSweeper`) is added to the `ClockLine` (on line 87 of labscript/core.py). This occurs in the constructor of the `AD9959DDSSweeper`, at which time `child_devices` is still empty.
If I am not mistaken, the `clock_limit` will still be applied, which could restrict the frequency of the `ClockLine` unnecessarily.

Reply:
Hmm, there are a lot of calls to `clock_limit`. Perhaps the better solution is to have a `calculate_clock_limit` function at the beginning of generate code (or nearby); after devices are fixed, but before change times are being handled.

Reply:
I suppose we are working with a fairly unusual situation (a timed device that has no timing but is still programmed). My only semi-coherent thought to add to this is: Is there some default value (like `None`) that we could set which gets interpreted by labscript as infinite? Presumably that is what was happening, at some level, when we forgot to set a limit at all initially. If so, we could use that default for Static-only code compilation.

Reply:
OK, I am understanding the issue slightly better now. I am not opposed to making `num_dyn_chans` a connection_table property constructor argument if it makes things easier. That said, we could also over-ride `clock_limit` as a property with a cached dynamic getter (to implement your second idea elegantly).